OOPLB3/Controller.cpp

```diff
@@ -37,21 +30,29 @@ void Controller::action(CommandReader &cur, Field &field) {
         case 'd':
             newPosition.second++;
             break;
+        case 'l':
+            notify(Message(LogType::GameState, "Game end"));
+            break;
+        default:
+            notify(Message(LogType::CriticalState, "Wrong command"));
     }
     newPosition.first =
             (newPosition.first % field.get_amountCellsY() + field.get_amountCellsY()) % field.get_amountCellsY();
     newPosition.second =
             (newPosition.second % field.get_amountCellsX() + field.get_amountCellsX()) % field.get_amountCellsX();
-    if (field.get_map()[newPosition.first][newPosition.second].get_characteristic() != STONE) {
+    if (field.get_map()[newPosition.first][newPosition.second].get_characteristic() != STONE &&
+        (field.getPlayerPosY() != newPosition.first || field.getPlayerPosX() != newPosition.second)) {
```

Comment:
Не бойся выносить повторяющиеся куски кода - особенно если они означают осмысленную сущность. Если возникает проблема - вероятно, что-то не так с типами, которые возвращает метод.
И кажется здесь `std::pair` лишний, как думаешь? По крайней мере если использовать отдельные переменные, то сложнее перепутать first & second - сразу видно, где `x`, а где `y`

```suggestion
    int x = 0;
    int y = 0;
    (...)
    const int fieldAmountY = field.get_amountCellsY();
    const int fieldAmountX = field.get_amountCellsX();
    y = (y % fieldAmountY + fieldAmountY ) % fieldAmountY;
    x = (x % fieldAmountX + fieldAmountX ) % fieldAmountX;
    (...)
    const auto& cell = field.get_map()[newPosition.first][newPosition.second]
    if (cell.get_characteristic() != STONE &&
        (field.getPlayerPosY() != y || field.getPlayerPosX() != x)) {
```